### ISSUE TITLE
Implement drop-shadow filter

### ISSUE THREAD

**Issue by jdm** [servo/servo#5187]
According to http://mxr.mozilla.org/servo/source/components/style/properties.mako.rs#2621 the drop-shadow filter is still unsupported.

@yodalee, would you be interested in this instead?

**Comment by jdm** [comment]
http://dev.w3.org/fxtf/filters/#funcdef-drop-shadow

**Comment by yodalee** [comment]
Well sure.

**Comment by yodalee** [comment]
I got a question that `drop-shadow` filter takes more parameters, but the current parser seems parse one parameter
https://github.com/servo/servo/blob/master/components/style/properties.mako.rs#L2744
So I think need to write a parse function for drop-shadow?

**Comment by jdm** [comment]
Yep!

**Comment by yodalee** [comment]
I think I need to add a value struct for `drop-shadow` in `style/values.rs`, like the `Angle` of filter `Hue-rotate`

**Comment by jdm** [comment]
The spec says that the values are parsed the same as the box-shadow property, so you should be able to reuse those.

**Comment by yodalee** [comment]
Yes, I just saw that, now trying to reuse it =w=.

**Comment by yodalee** [comment]
Sorry jdm, since I'm stuck on another issue, so I give this issue to @shinglyu. Is that OK?

**Comment by jdm** [comment]
Sure!

**Comment by shinglyu** [comment]
Got it, thanks

**Comment by shinglyu** [comment]
Are we talking about https://github.com/servo/servo/blob/master/components/style/properties.mako.rs#L2711 ? The line number changed after a few commit.

**Comment by shinglyu** [comment]
OK, @yodalee told me it's on line 2713

**Comment by shinglyu** [comment]
I have a very basic problem, I'm trying to use reuse the `SpecifiedBoxShadow`, but don't know how to `use` it in filter's longhand section.

**Comment by jdm** [comment]
Something like `style::properties::box_shadow::SpecifiedBoxShadow`, I think.

**Comment by yodalee** [comment]
In here:
https://github.com/servo/servo/blob/master/components/style/properties.mako.rs#L2910
add
`use super::box_shadow::SpecifiedBoxShadow;` works for me partially. There is still error message but not `unsolved name` error.

**Comment by shinglyu** [comment]
Ah yes, `super::` does the trick, thanks!

**Comment by shinglyu** [comment]
Just finish the parsing part, will work on gfx tomorrow.

**Comment by shinglyu** [comment]
Hmm... I am a little bit blocked on the gfx code. Most filters create some kind of Azure filter to do the trick, but the drop-shadow filter need to do a lot of masking, shifting and blurring. I can't see how the box-shadow code can fit into the `components/gfx/filters.rs` code.

**Comment by jdm** [comment]
I wonder if draw_surface_with_shadow is what we want here? I don't know much about this though; maybe @pcwalton does.

**Comment by jdm** [comment]
For what it's worth, the equivalent Gecko code appears to be http://mxr.mozilla.org/mozilla-central/source/gfx/src/FilterSupport.cpp#926 .

**Comment by shinglyu** [comment]
Hmm... My problem is that the `filters::create_filters()` needs to return an Azure filter, but box shadow needs multiple steps (which is already written in `paint_context::draw_box_shadow`). I don't know how to package the `draw_box_shadow()` into a `Azure::azure_hl::FilterType`

**Comment by shinglyu** [comment]
Oh I see! The Gecko implementation actually uses a FLOOD filter + 2 COMPOSITE filter, I guess I can do that too.

**Comment by shinglyu** [comment]
I was trying to translate the Gecko implementation into Rust, but I found some of the `FilterType`s are missing in rust-azure's `azure_hl.rs` bindings (but they are in c++) Is it because we only create rust binding for c++ when needed? Do I need to modify the rust-azure code to get this working? Or can I use any Azure's c++ code directly?

**Comment by shinglyu** [comment]
Just a note to myself:

```
23:06	SimonSapin	shinglyu: If there are features in libazure that you can’t use from rust-azure, it’s very possible that some bindings are missing
23:06	SimonSapin	and could be added
23:07	shinglyu	@SimonSapin: OK
23:08	shinglyu	SimonSapin: So what would be the suggested workflow?
23:09	shinglyu	SimonSapin: I fork rust-azure, make modifications
23:09	shinglyu	Change the servo dependency to my fork
23:09	shinglyu	Test it locally, then create PR for rust-azure, then create PR for servo, is that right?
23:09	SimonSapin	shinglyu: see http://doc.crates.io/guide.html#overriding-dependencies to use a local copy of a library in your Servo build
23:10	shinglyu	SimonSapin: Ah! Wonderful
23:10	SimonSapin	namely, have something like `paths = ["../rust-azure"]` in `.cargo/config` at the root of the Servo repository
23:11	SimonSapin	other than that, yes it’s the usual github workflow: fork, push a new branch, make a pull request
23:11	shinglyu	I see
23:13	SimonSapin	and once your rust-azure changes are in our repository, run `./mach update-carog -p azure --precise (the commit hash)` so make Servo use it without the path override
23:13	SimonSapin	that should be part of your Servo PR
```

**Comment by shinglyu** [comment]
Status update: 
The shadow part (to alpha => blur => offset) is almost done.
TODO: flood the shadow with specified color; Use composite filter to combine the original image with the shadow.

**Comment by shinglyu** [comment]
I am stuck, I want to use a COMPOSITE filter to combine the shadow I created with the original image, but I can't find the equivalent of 
```
composite->SetAttribute(ATT_COMPOSITE_OPERATOR, (uint32_t)COMPOSITE_OPERATOR_IN);
```
(http://mxr.mozilla.org/mozilla-central/source/gfx/src/FilterSupport.cpp#1034)
in rust-azure. 

I tried to create a binding in rust-azure, but the interface looks very different (see http://mxr.mozilla.org/servo/source/components/gfx/filters.rs#24 for example). Can anyone help me?

**Comment by shinglyu** [comment]
Oops, I was trying to say using `COMPOSITE_OPERATOR_OVER` in the previous comment.

**Comment by jdm** [comment]
Isn't https://github.com/servo/rust-azure/blob/master/src/azure_hl.rs#L1320 what you want? Presumably you need to create a new CompositeOperator type and implement the FilterAttribute trait for it (https://github.com/servo/rust-azure/blob/master/src/azure_hl.rs#L1521).

**Comment by shinglyu** [comment]
Thank you for pointing it out! Let me try.

**Comment by shinglyu** [comment]
I'm now implementing the [`FilterAttribute`](https://github.com/shinglyu/rust-azure/blob/dropshadow/src/azure_hl.rs#L1469) trait on [`CompositeAttribute`](https://github.com/shinglyu/rust-azure/blob/dropshadow/src/azure_hl.rs#L1406). I wonder if I should cast the enum `CompositeOperator` to `u32` [here](https://github.com/shinglyu/rust-azure/blob/dropshadow/src/azure_hl.rs#L1469), or just pass the enum and cast it later?

Also, I got the error 
```
   Compiling azure v0.1.0 (file:///home/shinglyu/workspace/servo/../rust-azure)
/home/shinglyu/workspace/servo/../rust-azure/src/azure_hl.rs:1465:60: 1465:65 error: cannot move out of borrowed content
/home/shinglyu/workspace/servo/../rust-azure/src/azure_hl.rs:1465         let CompositeAttribute::CompositeOperator(value) = *self;
                                                                                                                             ^~~~~
/home/shinglyu/workspace/servo/../rust-azure/src/azure_hl.rs:1465:51: 1465:56 note: attempting to move value to here
/home/shinglyu/workspace/servo/../rust-azure/src/azure_hl.rs:1465         let CompositeAttribute::CompositeOperator(value) = *self;
                                                                                                                    ^~~~~
/home/shinglyu/workspace/servo/../rust-azure/src/azure_hl.rs:1465:51: 1465:56 help: to prevent the move, use `ref value` or `ref mut value` to capture value by reference
error: aborting due to previous error
Could not compile `azure`.
```
But I can't see why this does not compile, but [this one](https://github.com/shinglyu/rust-azure/blob/dropshadow/src/azure_hl.rs#L1562) (for `GaussianBlurAttribute`) does.

**Comment by jdm** [comment]
The difference is that CompositeOperator is not marked as Copy, while the AzFloat inside of GaussianBlurAttribute::StdDeviation is. If you derive Copy for the new enum it should work fine. And I casting the enum in the code you linked looks fine to me.

**Comment by shinglyu** [comment]
In libazure code, the Composite filter can take arbitrary number of inputs, so it uses a `IN_COMPOSITE_IN_START=0` as the first input ID, then the second uses `IN_COMPOSITE_IN_START+ 1` and so on. But in the rust-azure wrapper, we uses 

```
1358 impl FilterInputIndex for CompositeInputs {
1359     fn azure_index(&self) -> u32 {
1360         AZ_IN_COMPOSITE_IN_START
1361     }
1362 }
```
, which is a enum that can't be added with an int directly. 

I don't know what's the best way to do it? Simply cast it to int? or should I implement some helper function that adds an assigned number to `IN_COMPOSITE_IN_START`?

**Comment by jdm** [comment]
Probably just cast to an int.

**Comment by Adenilson** [comment]
@shinglyu are you planning to continue working on this? If not, can I step in?

**Comment by shinglyu** [comment]
@Adenilson I wish I can finish it, but I'm too occupied right now. I can hand over the half-done prototype to you. It's a little tricky because you need to modify `rust-azure` as well.

**Comment by shinglyu** [comment]
@larsbergstrom This is the bug I mentioned. I believe this will be impacted by WebRender? Could you point me to the WebRender code (or some existing patch) so I can study? Thank you.

**Comment by larsbergstrom** [comment]
@glennw Can you provide @shinglyu any pointers to relevant WebRender code (or close this, if it's already implemented)? Thanks!

**Comment by shinglyu** [comment]
Just to keep a note here: I'm currently implementing the CSS parser for Stylo. As the [MDN doc](https://developer.mozilla.org/zh-TW/docs/Web/CSS/filter#drop-shadow()_2) said, there _should_ be a `<spread-radius>` paramter, but most browsers don't implement it. I'll still implement that in case we support spread radius in Servo's layout and gfx later.

**Comment by shinglyu** [comment]
Wait! the filter spec says the `spread` parameter is not supported so far, see: https://drafts.fxtf.org/filters/#funcdef-filter-drop-shadow

**Comment by shinglyu** [comment]
CSS parser is done in #14218, gw has some suggestion for the webrender part:


> 06:03	gw	shinglyu: that sounds like a fairly involved task! But looking at how the box-shadow primitives work, with using the dynamic prim cache is probably a good reference point
> 06:04	shinglyu	gw: I believe I have to add a new function in blend.fs.glsl?
06:05	shinglyu	gw: And in the code handing the filter, I add a new branch and try to make it point to the shader code?
06:06	gw	shinglyu: yep, that's right. As a first implementation, you could probably just directly calculate the drop shadow in that ps_blend shader, however long term we will probably want it to work similarly to box shadow, where it draws a small segment of the box shadow, and then stretches that across where the pixels are the same (that would go in ps_blend). This
06:06	gw	is a performance thing, since evaluating the shadow is quite expensive in a shader. But doing it directly in ps_blend would be fine as an initial implementation

**Comment by mstange** [comment]
I'm not sure how the "stretch the middle part" optimization would work for the blur filter. For box-shadows, the shape of the input is known: it's just a rectangle, possibly with rounded corners, so you can easily compute the size and location of the stretchable middle part. But the CSS blur filter supports arbitrary inputs! Detecting a stretchable middle part in the general case could get quite tricky.
I wouldn't worry about doing that type of optimization for now.
